### PR TITLE
daemon: re-enable reexec

### PR DESCRIPTION
**Pull request by mvo5**

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -78,6 +92,22 @@ func ExecInCoreSnap() {
...
+	if err != nil {
+		logger.Noticef("cannot stat %q: %s", exe, err)
+		return
+	}
+	coreSnapSnapd, err := os.Stat(full)
+	if err != nil {
+		logger.Noticef("cannot stat %q: %s", full, err)
+		return
+	}
+	if currentSnapd.ModTime().After(coreSnapSnapd.ModTime()) {
```
Comment (zyga) [inline cmd/cmd.go]: Smart.

**Comment by mvo5** [comment]
@zyga Thanks for the review. Definitely needs more work for snap-confine. I like your idea of just calling the one inside /snap/{core,ubuntu-core}/current (but we need to think confinement).

As for the testing, this is tricky:
- we need to make sure we deal with refreshes/reverts/removals of the core snap when snapd runs from those core snaps. And we need to have automatic tests for that too.
- we need to test that the first snapd that supports re-exec will always run tests against the latest edge snap. if we every break a the deb-snapd re-execing into edge we need to find out automatically (break because of new files needed or something). Its just too easy to forget about re-exec and the limitations of it for packaging changes.

Alternatively we can brainstorm about alternative ideas:
- make /usr/bin/snap a downloader only on classic that just installs a snapd and core snap
(much more work obviously).

File: HACKING.md [hunk 2]
```diff
@@ -158,12 +158,12 @@ To test the `snapd` REST API daemon on a snappy system you need to
 transfer it to the snappy system and then run:
 
     sudo systemctl stop snapd.service snapd.socket
-    sudo /lib/systemd/systemd-activate -E SNAPD_DEBUG=3 -E SNAP_REEXEC=0 -E SNAPD_DEBUG_HTTP=3 -l /run/snapd.socket -l /run/snapd-snap.socket ./snapd
+    sudo /lib/systemd/systemd-activate -E SNAPD_DEBUG=3 -E SNAPD_DEBUG_HTTP=3 -l /run/snapd.socket -l /run/snapd-snap.socket ./snapd
```
Comment (pedronis) [inline HACKING.md]: I'm confused if we are turning on rexeec, wasn't the original code suggestion here correct fore development ?

File: spread.yaml [hunk 3]
```diff
@@ -9,7 +9,6 @@ environment:
     PATH: $GOPATH/bin:/snap/bin:$PATH:/usr/lib/go-1.6/bin
     TESTSLIB: $PROJECT_PATH/tests/lib
     SNAPPY_TESTING: 1
-    SNAP_REEXEC: 0
```
Comment (pedronis) [inline spread.yaml]: I also don't undestand this? shouldn't we now need to run all tests both with reexec and without? to test the current snapd from trunk

File: HACKING.md (hunk 2 above)
Comment (mvo5) [inline HACKING.md]: Yes, I should have explained this better. I think we do not need it for two reasons:
- cmd.go checks for /proc/self/exe and tries to find that path under the core snaps base dir - if running from a git checkout there will be no /snap/core/current/home/user/go/src/github.com... so it won't re-exec
- the code checks if the re-exec snapd is older and not re-exec if it is older

However I am happy to revert this change if it feels safer to have it.

File: spread.yaml (hunk 3 above)
Comment (mvo5) [inline spread.yaml]: We put both our freshly build `snapd` and `snap` into our core snap during the tests. This means that we test re-exec into our freshly-build snapd. Do you think I should add a comment into the prepare.sh to make this clearer?

Ideally we would run all the tests twice, not sure if we want to bear the cost. Maybe we can have a spread-cron test with that?

File: HACKING.md (hunk 2 above)
Comment (pedronis) [inline HACKING.md]: I see, it's ok,  mostly because of point 1,  I can imagine scenario when one is trying to run an old locally built snapd, but this reminds me what I was discussing with Chipaca, that is we should log the version when we start snapd (it would help debugging in many situations)

File: cmd/cmd.go [hunk 4]
```diff
@@ -47,6 +47,19 @@ const oldCore = "/snap/ubuntu-core/current"
 // old ubuntu-core snaps older than this aren't suitable targets for re-execage
 const minOldRevno = 126
 
+func shouldReexec(envKey string) bool {
```
Comment (chipaca) [inline cmd/cmd.go]: this is `osutil.GetenvBool(envKey, true)` (with #2621)

Comment (mvo5) [inline cmd/cmd.go]: :-D

Comment (zyga) [inline cmd/cmd.go]: IMHO the function should probe the environment internally and just have a signature `func shouldReExec() bool`  but feel free to do this later.

File: spread.yaml (hunk 3 above)
Comment (zyga) [inline spread.yaml]: I think that a comment and spread-cron tasks are sensible.

I feel somewhat uneasy about releasing this tonight without good confidence that it doesn't explode.

File: spread.yaml [hunk 5]
```diff
@@ -170,20 +169,6 @@ prepare: |
         exit 1
     fi
 
-    # Disable burst limit so resetting the state quickly doesn't create problems.
```
Comment (zyga) [inline spread.yaml]: Why are we removing this?

File: tests/lib/prepare.sh [hunk 6]
```diff
@@ -14,29 +14,34 @@ update_core_snap_with_snap_exec_snapctl() {
     snap="$(mount | grep " $core" | awk '{print $1}')"
     umount --verbose "$core"
 
-    # Now unpack the core, inject the new snap-exec and snapctl into it, and
-    # repack it.
+    # Now unpack the core, inject the new snap-exec/snapctl into it
     unsquashfs "$snap"
     cp /usr/lib/snapd/snap-exec squashfs-root/usr/lib/snapd/
     cp /usr/bin/snapctl squashfs-root/usr/bin/
+    # also add snap/snapd because we re-exec by default.
```
Comment (zyga) [inline tests/lib/prepare.sh]: I'll say this again just for completeness. I think that ultimately reexec should be fully symmetric.

If it is enabled we always take snapd/snap/snap-confine/snapctl/snap-exec from the core snap, if it is more recent than from the pacakge. If it is disabled we always take snapd/snap-confine/snapctl/snap-exec from the classic package. Why? Because this ensures that the set is matched. This especially matters for snapctl and snap-exec as we learned in the past.

I think that re-packaging aside. We should use snap-confine to ensure this policy is used.

File: tests/lib/prepare.sh [hunk 7]
```diff
@@ -55,6 +60,26 @@ prepare_classic() {
         exit 1
     fi
 
+    # Disable burst limit so resetting the state quickly doesn't create
```
Comment (zyga) [inline tests/lib/prepare.sh]: Aha, I guess this answers my earlier question.

File: tests/main/writable-areas/task.yaml [hunk 8]
```diff
@@ -1,5 +1,9 @@
 summary: Check that snap apps and services can write to writable areas.
 
+environment:
+    SNAP_REEXEC/norexec: 0
+    SNAP_REEXEC/withreexec: 1
+
```
Comment (zyga) [inline tests/main/writable-areas/task.yaml]: Can we please have a smoke test that ensures that hello world (or similar) works under reexec.

File: tests/main/snapd-reexec/task.yaml [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+systems: [-ubuntu-core-16-64, -ubuntu-core-16-arm-64, -ubuntu-core-16-arm-32]
+
+environment:
+    SNAPD_CONF: /etc/systemd/system/snapd.service.d/local.conf 
+
+execute: |
+    if [ "${SNAP_REEXEC:-}" = "0" ]; then
+        echo "skipping test when SNAP_REEXEC is disabled"
+        exit 0
```
Comment (zyga) [inline tests/main/snapd-reexec/task.yaml]: gee, I wish we had SKIP in spread

File: tests/main/snapd-reexec/task.yaml [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    fi
+
+    echo "Ensure we re-exec by default"
```
Comment (zyga) [inline tests/main/snapd-reexec/task.yaml]: Can `snap --version` say something that re-exec has occurred? (or maybe this will be obvious from versions).

Can we check for that?

File: tests/main/snapd-reexec/task.yaml [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    snap list
+    journalctl | MATCH "DEBUG: restarting into"
+
+    echo "Ensure that we do not re-exec into older versions"
+    systemctl stop snapd.service snapd.socket
+    echo "mount something older than our freshly build snapd"
+    mount --bind /bin/true /snap/core/current/usr/lib/snapd/snapd
+    systemctl start snapd.service snapd.socket
+    snap list
+    journalctl | MATCH "not restarting into.*older than"
```
Comment (zyga) [inline tests/main/snapd-reexec/task.yaml]: 😍

File: tests/main/snapd-reexec/task.yaml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    #!/bin/sh
+    echo "from the core snap"
+    exit 1
+    EOF
+    chmod +x /tmp/broken-snapd
+    mount --bind /tmp/broken-snapd /snap/core/current/usr/lib/snapd/snapd
+    systemctl daemon-reload
+    systemctl start snapd.service snapd.socket
+    echo "Ensure that snap list works normally"
+    echo "(i.e. the snapd from the core image is not run)"
```
Comment (zyga) [inline tests/main/snapd-reexec/task.yaml]: Sorry for being so dumb, how does snapd self-defend against this?

File: tests/main/snapd-reexec/task.yaml [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    snap list | MATCH core
+
+    echo "Revert back to normal"
```
Comment (zyga) [inline tests/main/snapd-reexec/task.yaml]: This could be a small function to make the test easier to follow (not looking for paranoia typos)

File: tests/main/snapd-reexec/task.yaml [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    mount --bind /tmp/broken-snapd /snap/core/current/usr/bin/snap
+    snap list|MATCH "from the core snap"
+    SNAP_REEXEC=0 snap list|MATCH "core"
+    umount /snap/core/current/usr/bin/snap
+
+    echo "Ensure a core refresh restart snapd"
+    . $TESTSLIB/names.sh
+    prev_core=$(snap list | awk "/^${core_name} / {print(\$3)}")
+    snap install --dangerous /var/lib/snapd/snaps/${core_name}_${prev_core}.snap
+    journalctl | MATCH "Requested daemon restart"
```
Comment (zyga) [inline tests/main/snapd-reexec/task.yaml]: Curious, how does this work? I assumed this only happens on core.

File: spread.yaml (hunk 5 above)
Comment (mvo5) [inline spread.yaml]: It's moved only, it appears again in prepare.sh a but further down.

File: tests/lib/prepare.sh (hunk 6 above)
Comment (mvo5) [inline tests/lib/prepare.sh]: Indeed. From the current re-exec snap-confine is missing. Mostly because we need to think about confinement for that (the current approach won't work). However if we do pick which snap-confine to use in snapd itself we can easily get this fix in via a new core snap (as the snapd from the core snap will select which snap-confine to run).

As a side-note. Running snap-exec/snapctl from classic is right now hard because we are already inside the core snap at this point. But we could write code that somehow bind mounts them across (which I think has its own problems because how do we know if we must bind mount more than just the binary).

I think ideally we would have a single "snapd" snap that contains all this code and we don't re-exec at all but use the snap everywhere. On classic the deb of snapd would just contain a /usr/bin/snap "loader" that downlaods the snapd (or core) snap and then runs it. but obviously quite a bit more work and we need to think about transitions etc.

File: tests/main/writable-areas/task.yaml (hunk 8 above)
Comment (mvo5) [inline tests/main/writable-areas/task.yaml]: All tests are designed to run with re-exec. We put snap/snapd into the core snap so that all tests use them. There is also e.g. "install-sideload" that will run with both re-exec on and off and that installs and runs  hello-world (test-snapd-tools to be exact).

File: tests/main/snapd-reexec/task.yaml (hunk 10 above)
Comment (mvo5) [inline tests/main/snapd-reexec/task.yaml]: We currently write into syslog that we restart and into what path. Do you think that is sufficient?

File: tests/main/snapd-reexec/task.yaml (hunk 12 above)
Comment (mvo5) [inline tests/main/snapd-reexec/task.yaml]: this test just tests that SNAP_REEXEC is honored. Unfortunately we do not detect a broken snapd in the core snap. When it looks valid we just re-exec into it.

File: tests/main/snapd-reexec/task.yaml (hunk 14 above)
Comment (mvo5) [inline tests/main/snapd-reexec/task.yaml]: We have the same mechanism that triggers a reboot on core trigger a restart of snapd in classic :) and now a spread test for it too.

**Comment by mvo5** [comment]
Thanks for the review @zyga ! I understand your concerns, its a slightly risky feature. The added tests (and the changes to the tests) hopefully make it safe but its stil risky. I think in the long run we want to move to a snap for snapd but that is still some way away it seems.

File: spread.yaml (hunk 3 above)
Comment (pedronis) [inline spread.yaml]: after further thinking, I think is very strange to run our main tests for master using reexec on... we really need a 2nd suite I think or a spread-cron task I think.

I mean our main suite should run SNAP_REEXEC=0 because it's about testing the code we are writing atm

File: tests/main/enable-disable/task.yaml [hunk 15]
```diff
@@ -1,5 +1,9 @@
 summary: Check that enable/disable works
 
+environment:
+    SNAP_REEXEC/norexec: 0
```
Comment (pedronis) [inline tests/main/enable-disable/task.yaml]: typo here: noreexec ?

File: tests/main/install-remove-multi/task.yaml [hunk 16]
```diff
@@ -1,5 +1,9 @@
 summary: Check that install/remove of multiple snaps works
 
+environment:
+    SNAP_REEXEC/norexec: 0
+    SNAP_REEXEC/withreexec: 1
+
```
Comment (pedronis) [inline tests/main/install-remove-multi/task.yaml]: probably ok to drop this one

File: tests/main/install-store-laaaarge/task.yaml [hunk 17]
```diff
@@ -1,5 +1,9 @@
 summary: snap install a large snap from the store (bigger than tmpfs)
 
+environment:
+    SNAP_REEXEC/norexec: 0
+    SNAP_REEXEC/withreexec: 1
+
```
Comment (pedronis) [inline tests/main/install-store-laaaarge/task.yaml]: probably ok to drop this one

File: tests/main/enable-disable/task.yaml (hunk 15 above)
Comment (pedronis) [inline tests/main/enable-disable/task.yaml]: also maybe we alway want a comment  # test the no-reexec code path

File: spread.yaml [hunk 18]
```diff
@@ -9,7 +9,10 @@ environment:
     PATH: $GOPATH/bin:/snap/bin:$PATH:/usr/lib/go-1.6/bin
     TESTSLIB: $PROJECT_PATH/tests/lib
     SNAPPY_TESTING: 1
-    SNAP_REEXEC: 0
+    # we run the entire suite with re-exec on and modify the core
```
Comment (pedronis) [inline spread.yaml]: ... re-exec on (the default) ...

File: tests/main/snapd-reexec/task.yaml [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    systemctl stop snapd.service snapd.socket
+    echo "mount something older than our freshly build snapd"
+    mount --bind /bin/true /snap/core/current/usr/lib/snapd/snapd
+    systemctl start snapd.service snapd.socket
+    snap list
+    journalctl | MATCH "not restarting into.*older than"
+
+    echo "Revert back to normal"
+    systemctl stop snapd.service snapd.socket
+    umount /snap/core/current/usr/lib/snapd/snapd
```
Comment (pedronis) [inline tests/main/snapd-reexec/task.yaml]: we probably somehow want to do a best effort at this umount also in  restore stanza

bit worried than when this test fail it will leave a strange state around

File: tests/main/snapd-reexec/task.yaml [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    systemctl daemon-reload
+    systemctl start snapd.service snapd.socket
+    echo "Ensure that snap list works normally"
+    echo "(i.e. the snapd from the core image is not run)"
+    snap list | MATCH core
+
+    echo "Revert back to normal"
+    systemctl stop snapd.service snapd.socket
+    umount /snap/core/current/usr/lib/snapd/snapd
+    rm -f /etc/systemd/system/snapd.service.d/no-reexec.conf
```
Comment (pedronis) [inline tests/main/snapd-reexec/task.yaml]: same for this rm -f

File: tests/main/snapd-reexec/task.yaml [hunk 21]
```diff
@@ -0,0 +1,78 @@
+summary: Test that snapd rexecs itself into core
```
Comment (pedronis) [inline tests/main/snapd-reexec/task.yaml]: typo reexec

File: spread.yaml (hunk 3 above)
Comment (pedronis) [inline spread.yaml]: this was clarified now, we run with reexec but with a freshened core

File: tests/main/enable-disable/task.yaml [hunk 22]
```diff
@@ -1,5 +1,9 @@
 summary: Check that enable/disable works
 
+environment:
+    SNAP_REEXEC/noreexec: 0
```
Comment (pedronis) [inline tests/main/enable-disable/task.yaml]: this one is missing the comment, also not sure we need (as mentioned in tg)

File: spread.yaml [hunk 23]
```diff
@@ -9,7 +9,10 @@ environment:
     PATH: $GOPATH/bin:/snap/bin:$PATH:/usr/lib/go-1.6/bin
     TESTSLIB: $PROJECT_PATH/tests/lib
     SNAPPY_TESTING: 1
-    SNAP_REEXEC: 0
+    # we run the entire suite with re-exec on (the default) and modify
+    # the core snap so that it contains our new code.  So we run new
+    # snapd from the deb that re-execs into new snapd in core.  To
+    # test purely from the deb, the "SNAP_REEXEC: 0" here
```
Comment (fgimenez) [inline spread.yaml]: We could add here an environment variable so that we can parameterize the modification of the core snap later, something like:
```
MODIFY_CORE_SNAP_FOR_REXEC: "$(HOST: echo ${SPREAD_MODIFY_CORE_SNAP_FOR_REXEC:-1})"
```

This way we could control the modification of the core snap from the host, modifying it by default.

File: tests/lib/prepare.sh [hunk 24]
```diff
@@ -74,7 +101,7 @@ prepare_classic() {
 
         systemctl stop snapd.service snapd.socket
 
-        update_core_snap_with_snap_exec_snapctl
+        update_core_snap_for_classic_reexec
```
Comment (fgimenez) [inline tests/lib/prepare.sh]: And we could only call `update_core_snap_for_classic_reexec` only when `MODIFY_CORE_SNAP_FOR_REXEC` is set to 1.

File: tests/lib/prepare.sh [hunk 25]
```diff
@@ -5,6 +5,12 @@ set -eux
 . $TESTSLIB/apt.sh
 
 update_core_snap_for_classic_reexec() {
+    # it is possible to disable this to test that snapd (the deb) works
+    # fine with whatever is in the core snap
+    if [ "$MODIFY_CORE_SNAP_FOR_REEXEC" != "1" ]; then
+        return
```
Comment (pedronis) [inline tests/lib/prepare.sh]: should we echo something here ?

File: spread.yaml [hunk 26]
```diff
@@ -9,7 +9,12 @@ environment:
     PATH: $GOPATH/bin:/snap/bin:$PATH:/usr/lib/go-1.6/bin
     TESTSLIB: $PROJECT_PATH/tests/lib
     SNAPPY_TESTING: 1
-    SNAP_REEXEC: 0
+    # we run the entire suite with re-exec on (the default) and modify
+    # the core snap so that it contains our new code.  So we run new
+    # snapd from the deb that re-execs into new snapd in core.  To
+    # test purely from the deb, set "export SPREAD_SNAP_REEXEC=0"
+    SNAP_REEXEC: "$(HOST: echo ${SPREAD_SNAP_REEXEC:-})"
```
Comment (pedronis) [inline spread.yaml]: we should maybe `echo running with SNAPREXEC: "$SNAP_REEXEC"` around the top of the main prepare stanza ?

File: spread.yaml (hunk 3 above)
Comment (niemeyer) [inline spread.yaml]: But is there a clear benefit in not being explicit here, given we have a few tests explicitly checking that the reexec functionality works properly? It feels somewhat easy to do a major mistake and not test the snapd just built at all, and release something absurdly broken.

File: tests/main/install-errors/task.yaml [hunk 27]
```diff
@@ -4,6 +4,10 @@ systems: [-ubuntu-core-16-64, -ubuntu-core-16-arm-64, -ubuntu-core-16-arm-32]
 
 environment:
     SNAP_NAME: test-snapd-tools
+    # Ensure that running purely from the deb (without re-exec) works
+    # correctly
+    SNAP_REEXEC/noreexec: 0
+    SNAP_REEXEC/withreexec: 1
```
Comment (niemeyer) [inline tests/main/install-errors/task.yaml]: Suggestion for here and below:
```
SNAP_REEXEC/reexec0: 0
SNAP_REEXEC/reexec1: 1

File: debian/snapd.service [hunk 28]
```diff
@@ -4,6 +4,8 @@ Requires=snapd.socket
 
 [Service]
 ExecStart=/usr/lib/snapd/snapd
+# add "SNAP_REEXEC=0" to /etc/environment to disable the behaviour of
+# snapd to re-exec itself into the snapd of the core snap
```
Comment (mvo5) [inline debian/snapd.service]: So your suggestion is to just remove the comment?

File: cmd/cmd.go (hunk 1 above)
Comment (mvo5) [inline cmd/cmd.go]: This is meant as a cheap way to ensure that people get the latest snapd. the problem is of course that if the deb is rebuild it won't reexec into core anymore. I'm not sure that this is actually a problem as usually when people rebuild (or the distro rebuilds) there is a reason (like bugfix in the toolchain) so it seems sensible to use the newer version still. However if you think the behaviour is undesirable I'm happy to remove it again. Should it just unconditionally re-exec? Is that your suggestion?

File: tests/main/install-errors/task.yaml (hunk 27 above)
Comment (mvo5) [inline tests/main/install-errors/task.yaml]: Thanks! Updated the code.

File: spread.yaml (hunk 3 above)
Comment (mvo5) [inline spread.yaml]: I see the benefit that this ensures that everything works correctly when we do re-exec (which will be the default for most people that it does re-exec). We can test with SNAP_REEXEC=0 here of course, I think this would actually test slightly less because this way we don't know that everything still works when it does the re-exec operation.

File: spread.yaml (hunk 26 above)
Comment (mvo5) [inline spread.yaml]: Thanks! Added that.

File: tests/lib/prepare.sh (hunk 24 above)
Comment (mvo5) [inline tests/lib/prepare.sh]: this is done now.

File: cmd/cmd.go (hunk 1 above)
Comment (niemeyer) [inline cmd/cmd.go]: That means it will arbitrarily move back and forth depending on who touches last, completely independent of the content. Feels very dangerous.. easy to brick. Revision seems much more stable and reliable. Do I miss your intention?

Comment (pedronis) [inline cmd/cmd.go]: nothing that unconditionally re-exec opens again a scenario where we need even more test cases

File: spread.yaml (hunk 3 above)
Comment (niemeyer) [inline spread.yaml]: Okay, sounds good.

File: cmd/cmd.go (hunk 1 above)
Comment (mvo5) [inline cmd/cmd.go]: As discussed this will be improved and we will use a /usr/lib/snapd/info file with "VERSION=$VER" in it.

File: cmd/cmd.go [hunk 29]
```diff
@@ -67,31 +65,34 @@ func ExecInCoreSnap() {
...
 
-	// ensure we do not re-exec into an older version of snapd
-	currentSnapd, err := os.Stat(exe)
-	if err != nil {
-		logger.Noticef("cannot stat %q: %s", exe, err)
-		return
+	// ensure we do not re-exec into an older version of snapd, look
+	// for info file and ignore version of core that do not yet have
+	// it
+	fullInfo := filepath.Join(newCore, "/usr/lib/snapd/info")
```
Comment (pedronis) [inline cmd/cmd.go]: can't we remember which one we used for finding full inside a core, instead of trying again this way here?

File: cmd/sort.go [hunk 30]
```diff
@@ -0,0 +1,198 @@
...
+		return res
+	}
+
+	// the subversion revision behind the "-"
+	revA := strings.Split(va, "-")[1]
+	revB := strings.Split(vb, "-")[1]
+	return compareSubversion(revA, revB)
+}
+
+// ByVersion provides a sort interface
```
Comment (pedronis) [inline cmd/sort.go]: we can drop this stuff, no?

File: cmd/sort.go [hunk 31]
```diff
@@ -0,0 +1,198 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cmd
```
Comment (pedronis) [inline cmd/sort.go]: wondering if this one should go in strutil/version.go

Comment (mvo5) [inline cmd/sort.go]: :+1: I like that. will do

File: cmd/cmd.go [hunk 32]
```diff
@@ -66,17 +65,36 @@ func ExecInCoreSnap() {
...
+		if !osutil.FileExists(fullInfo) {
 			return
 		}
 	}
+	content, err := ioutil.ReadFile(fullInfo)
+	if err != nil {
+		logger.Noticef("cannot read info file %q: %s", fullInfo, err)
+		return
+	}
+	ver := regexp.MustCompile("(?ms)^VERSION=(.*)\n").FindStringSubmatch(string(content))
```
Comment (chipaca) [inline cmd/cmd.go]: You want `m`, sure, so the `^` and `$` match in any line of the file, but not `s`: you don't want that `.*` to eat everything to the end of the file.
Also, instead of `\n`, just use `$` (so that you don't *need* to have a newline at EOF).

Comment (mvo5) [inline cmd/cmd.go]: Indeed, thanks for this. Fixed in git.

File: cmd/sort.go (hunk 30 above)
Comment (mvo5) [inline cmd/sort.go]: Yes, I removed it.

File: cmd/cmd.go (hunk 29 above)
Comment (mvo5) [inline cmd/cmd.go]: Good point, I changed the code accordingly.
